### ISSUE TITLE
Floating point exception in gpuR test against pocl

### ISSUE THREAD

**Issue by znmeb** [cdeterman/gpuR#78]
How to reproduce:

1. Install pocl OpenCL 2.0 pocl 0.14-pre, LLVM 3.9.0
2. Install gpuR 1.2.2
3. Start R and type `devtools::test()` in the GitHub directory

```
> devtools::test()
Loading gpuR
Loading required package: testthat
Creating a generic function for ‘eigen’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘%o%’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘nrow’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘ncol’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘crossprod’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘tcrossprod’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘cov’ from package ‘stats’ in package ‘gpuR’
Creating a generic function for ‘colSums’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘rowSums’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘colMeans’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘rowMeans’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘dist’ from package ‘stats’ in package ‘gpuR’
Creating a generic function for ‘diag’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘diag<-’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘qr.R’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘qr.Q’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘typeof’ from package ‘base’ in package ‘gpuR’
Creating a generic function for ‘colnames<-’ from package ‘base’ in package ‘gpuR’
Number of platforms: 1
- platform: The pocl project: OpenCL 2.0 pocl 0.14-pre, LLVM 3.9.0
  - gpu index: 0
    - pthread-AMD FX(tm)-8350 Eight-Core Processor
checked all devices
completed initialization
gpuR 1.2.2
Testing gpuR
CPU deepcopy: ........................................
CPU gpuMatrix algebra: ..................................1...............................................................................2...................
CPU gpuMatrix chol decomposition: ......
CPU gpuMatrix classes: .............................................
CPU gpuMatrix Correlations: ....
CPU gpuMatrix Distance Computations: ..........................
CPU gpuMatrix eigen decomposition: ......
CPU gpuMatrix math operations: .......................345...678..............
CPU gpuMatrix qr decomposition: ........
CPU gpuMatrix Row and Column Methods: ........................................................
CPU gpuMatrix solve: ........................
CPU gpuMatrix svd decomposition: ..........
CPU gpuMatrix Utility Functions: ............................................................
CPU gpuMatrixBlock algebra: .................................9............................................a...........
CPU gpuVector algebra: .....................................b......................................c.........
CPU gpuVector classes: .................................
CPU gpuVector math operations: .......................def...FFF..............
CPU gpuVector Utility Functions: ...................
CPU Inplace Operations: ........
CPU Ordering Methods: ....
CPU vclVector shared memory: ....
CPU vclMatrix algebra: ................................F.....................................Floating point exception (core dumped)
```
If I can get `gdb` to behave I'll grab a backtrace if you want one.

**Comment by cdeterman** [comment]
@znmeb Sure, the `gdb` backtrace could be useful.  However, it may also be useful if we localize this a bit more.  Can you just run `devtools::test_file('tests/testthat/test_cpu_gpuMatrix_algebra.R')`?  Perhaps we can get a little more information beyond the core dump.

**Comment by znmeb** [comment]
```
> devtools::test_file('tests/testthat/test_cpu_gpuMatrix_algebra.R')
Error: 'test_file' is not an exported object from 'namespace:devtools'
> testthat::test_file('tests/testthat/test_cpu_gpuMatrix_algebra.R')
Number of platforms: 1
- platform: The pocl project: OpenCL 2.0 pocl 0.14-pre, LLVM 3.9.0
  - gpu index: 0
    - pthread-AMD FX(tm)-8350 Eight-Core Processor
checked all devices
completed initialization
gpuR 1.2.1

Attaching package: ‘gpuR’

The following objects are masked from ‘package:base’:

    colnames, svd

CPU gpuMatrix algebra: ..................................1...............................................................................2...................

Failed -------------------------------------------------------------------------
1. Failure: CPU gpuMatrix Single Precision Matrix Element-Wise Power (@test_cpu_gpuMatrix_algebra.R#217) 
fgpuC[, ] not equal to `C`.
10/16 mismatches (average diff: NaN)
[2]  -1.811 - NaN == NaN
[4]  -0.597 - NaN == NaN
[5]  -1.037 - NaN == NaN
[6]   1.099 - NaN == NaN
[7]   0.992 - NaN == NaN
[8]   1.128 - NaN == NaN
[9]  -1.386 - NaN == NaN
[10]  0.897 - NaN == NaN
[13] -1.169 - NaN == NaN
...
float matrix elements not equivalent

2. Failure: CPU gpuMatrix Double Precision Matrix Element-Wise Power (@test_cpu_gpuMatrix_algebra.R#736) 
fgpuC[, ] not equal to `C`.
10/16 mismatches (average diff: NaN)
[2]  -1.811 - NaN == NaN
[4]  -0.597 - NaN == NaN
[5]  -1.037 - NaN == NaN
[6]   1.099 - NaN == NaN
[7]   0.992 - NaN == NaN
[8]   1.128 - NaN == NaN
[9]  -1.386 - NaN == NaN
[10]  0.897 - NaN == NaN
[13] -1.169 - NaN == NaN
...
double matrix elements not equivalent

DONE ===========================================================================
> 
```

**Comment by cdeterman** [comment]
@znmeb I wanted to update this issue to confirm that I am able to reproduce the problem.  When I have tried to switch over to using pocl for you Travis CI unit testing I noticed that the tests were failing.  This appears to be a `pocl` platform issue as it works locally on my machine with is using the Intel OpenCL platform.  I have opened an issue with the `pocl` developers and hopefully will hear back soon.
